src/coreclr/jit/assertionprop.cpp

```diff
@@ -1200,8 +1200,18 @@ AssertionIndex Compiler::optCreateAssertion(GenTree*         op1,
                     }
                     else
                     {
-                        noway_assert(op2->gtOper == GT_CNS_DBL);
                         /* If we have an NaN value then don't record it */
+                        noway_assert(op2->gtOper == GT_CNS_DBL);
+#ifdef TARGET_RISCV64
+                        if (op2->TypeGet() == TYP_FLOAT)
+                        {
+                            double f64Cns = op2->AsDblCon()->DconValue();
+                            if (_isnan(*reinterpret_cast<float*>(&f64Cns)))
+                            {
+                                goto DONE_ASSERTION;
+                            }
+                        }
+#endif // TARGET_RISCV64
```

Comment:
Why is this change necessary? _isnan should not care about payload bits.

Reply:
Sure. I will make a new helper. Thank you. I want to know proper location and name.

Because when a float value is converted to `Double` with preserving payload and saved in `DconValue`, the saved `Double` value is not 'NaN' in `Double` type.
For example, `c` can be assumed as `f64Cns` in the change.
```
    int a = -1;
    double c = *reinterpret_cast<double*>(&a);

    float d = *reinterpret_cast<float*>(&c);
    float e = (float)c;
    fprintf(stderr, "%lf %f %f %d %d %d\n", c, d, e, isnan(c), isnan(d), isnan(e));
```
0.000000 nan 0.000000 0 1 0
Only the `*reinterpret_cast<float*>(&c)` is `NaN` in the example.
If I misunderstand your comment, please let me know.
Thank you!

Reply:
Conversion to a double by reinterpreting a 32 bit int is not correct. The right way to do it is something that constructs the uint64_t bits directly by ORing in the desired payload bits in the right place and returns `UInt64BitsToDouble` from that. It should mimic what our other hosts do for float -> double conversions. We should not be trying to store a 32 bit float in the 64 bit double field.

I currently do not have access to a PC, but @tannergooding may be able to help with the exact right way to do the conversion that retains the payload bits on the proper way.

Reply:
The PR updates codes to use `fmv` (preserving payload) or `fcvt` (not preserving payload) in case of float is `NaN` or not.
I will wait for the exact solution. Thank you so much!!!

Reply:
I think there are a few possible options. For the JIT side, all of them have the issue that the general codebase would need to be "audited" to ensure that all float->double handling is done correctly. We're not consistent about this today.

## Update GenTreeDblCon to be more like GenTreeIcon

You can update `GenTreeDblCon` to have a union of `gtDconVal` and `gtFconVal`. It would then expose `float FconValue()` and `void SetFconValue(float value)` and assert that `Dcon` is only called for `TYP_DOUBLE` and `Fcon` is only called for `TYP_FLOAT

This will ensure that the values aren't ever accessed incorrectly and avoids the upcast/downcast issue entirely for the JIT. It does not resolve the issue for managed code and will likely have some minor TP hit due to needing to add branching to various callsites in the JIT.

## Expose a payload preserving conversion helper

Just as it says, define a helper API that does a payload preserving conversion. This would be a software based conversion and would simply ensure the significand bits are copied up and truncated down, just as happens for other platforms.

This helper would be used by both the JIT and managed and ensures managed code also gets the right handling. This would incur a branch check per conversion on RISC-V.

## Explicitly utilize `FMV.n.X`/`FMV.X.n`

These instructions allow storing `n-bits` in the larger `f` register, preserving the payload. The value stored is a `NaN boxed value`.

One could utilize this to more efficiently store the data in `gtDconVal`, but it has many of the same issues/considerations where it needs to be done everywhere and consistently; otherwise subtle bugs could creep in.

Reply:
I'd suggest to introduce the payload preserving conversion helper for now and use it in the cases found in this PR. I think that's going to be the least invasive fix. We can consider the more complete/invasive fixes as part of .NET 9.

Reply:
@jakobbotsch @tannergooding Thank you so much for the comments.
I made a conversion helper and updated all. However, I don't know what is the right way to convert, so I just retained the conversion method as it is. Could you please give the right way what you mentioned?
> Conversion to a double by reinterpreting a 32 bit int is not correct. The right way to do it is something that constructs the uint64_t bits directly by ORing in the desired payload bits in the right place and returns UInt64BitsToDouble from that.

Reply:
> @jakobbotsch @tannergooding Thank you so much for the comments. I made a conversion helper and updated all. However, I don't know what is the right way to convert, so I just retained the conversion method as it is. Could you please give the right way what you mentioned?
> 
> > Conversion to a double by reinterpreting a 32 bit int is not correct. The right way to do it is something that constructs the uint64_t bits directly by ORing in the desired payload bits in the right place and returns UInt64BitsToDouble from that.

Could you take a look? Thank you so much.

Reply:
@clamp03 The conversion functions should look something like:
```cpp
//------------------------------------------------------------------------
// convertToDouble: Convert a single to a double with platform independent
// preservation of payload bits.
//
// Arguments:
//   f - the single
//
// Return Value:
//   A double.
//
// Remarks:
//   All our host platforms except for RISCV-64 will preserve payload bits of
//   NaNs. This function implements the conversion in software for RISCV-64 to
//   mimic other platforms.
//
double FloatingPointUtils::convertToDouble(float f)
{
#ifdef HOST_RISCV64
    if (f == f)
    {
        return f;
    }

    uint32_t bits = BitOperations::SingleToUInt32Bits(f);
    uint32_t payload = bits & ((1u << 23) - 1);
    uint64_t newBits = ((uint64_t)(bits >> 31) << 63) | 0x7FF8'0000'0000'0000ul | ((uint64_t)payload << 29);
    return BitOperations::UInt64BitsToDouble(newBits);
#else
    return f;
#endif
}

//------------------------------------------------------------------------
// convertToSingle: Convert a double to a single with platform independent
// preservation of payload bits.
//
// Arguments:
//   d - the double
//
// Return Value:
//   A float.
//
// Remarks:
//   All our host platforms except for RISCV-64 will preserve payload bits of
//   NaNs. This function implements the conversion in software for RISCV-64 to
//   mimic other platforms.
//
float FloatingPointUtils::convertToSingle(double d)
{
#ifdef HOST_RISCV64
    if (d == d)
    {
        return (float)d;
    }

    uint64_t bits = BitOperations::DoubleToUInt64Bits(d);
    uint32_t newPayload = (uint32_t)((bits >> 29) & ((1u << 23) - 1));
    uint32_t newBits = ((uint32_t)(bits >> 63) << 31) | 0x7F80'0000u | newPayload;
    return BitOperations::UInt32BitsToSingle(newBits);
#else
    return (float)d;
#endif
}
```

You should change the JIT to use this conversion function in the places you have identified.

Reply:
Awesome!!! I didn't think so. Thank you so much.